google-cloud-spanner/src/main/java/com/google/cloud/spanner/TransactionRunnerImpl.java

```diff
@@ -462,11 +480,15 @@ TransactionSelector getTransactionSelector() {
             // The first statement of a transaction that gets here will be the one that includes
             // BeginTransaction with the statement. The others will be waiting on the
             // transactionIdFuture until an actual transactionId is available.
+            // If inlineCommit is true, the first statement should wait on the transactionIdFuture.
             if (transactionIdFuture == null) {
               transactionIdFuture = SettableApiFuture.create();
               if (trackTransactionStarter) {
                 transactionStarter = new Exception("Requesting new transaction");
               }
+              if (options.inlineCommit()) {
+                tx = transactionIdFuture;
```

Comment:
I don't quite understand how this would work. the `transactionIdFuture` has just been created above as a `SettableApiFuture` with no value. On line 511 we will now be waiting for this future to return a value, but no one will be setting a value for it. Or am I missing something here?

Reply:
About 20 lines below, there is a call to tx.get() with comment "wait for the transaction to come available..." I believe that call will set the value for it. No?

Reply:
No. That line of code will wait for 60 seconds for a transaction to come available when __a different statement__ has already been executed and included an inlined `BeginTransaction` option. An application can use the async API of the `TransactionRunner` to execute multiple statements (partly) in parallel. Consider the following example:

```java
// The following query will include the BeginTransaction option in the request.
var rs1 = transaction.executeQueryAsync("SELECT * FROM Foo");

// The following query must wait until the first query has returned a transaction ID before it can start.
var rs2 = transaction.executeQueryAsync("SELECT * FROM Bar");
```

The first query in the above example will initiate the transaction by inlining the BeginTransaction option. The second query would normally have been executed in parallel with the first query, but in this case it has to wait for the first one to return a transaction id. That is the `wait for the transaction to come available...` part does in the code.

The reason that this (probably) works is because it is dead code. The if-branch will only be entered `if (transactionIdFuture == null)`. When inline commit is enabled, this will never be null, because it does not use inline begin. That means that the first action of the transaction is executing a `BeginTransaction` RPC which will initialize the `transactionIdFuture`.

Short version: I'm pretty sure that you can (and should) remove this whole if-block.

Reply:
Indeed when using inlineCommit, transactionId cannot be empty. I've added a checkState() at the beginning of the "if (transactionId==null)" block.
This block cannot be removed because it is still in use when inlineCommit is not used.